resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeHelper.scala

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.yarn
+
+import java.lang.{Integer => JInteger, Long => JLong}
+import java.lang.reflect.InvocationTargetException
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import org.apache.hadoop.yarn.api.records.Resource
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils
+
+/**
+ * This helper class uses some of Hadoop 3 methods from the YARN API,
+ * so we need to use reflection to avoid compile error when building against Hadoop 2.x
+ */
+private object ResourceTypeHelper extends Logging {
+  private val AMOUNT_AND_UNIT_REGEX = "([0-9]+)([A-Za-z]*)".r
+  private val RESOURCE_TYPES_NOT_AVAILABLE_ERROR_MESSAGE =
+    "Ignoring updating resource with resource types because " +
+        "the version of YARN does not support it!"
+
+  def setResourceInfoFromResourceTypes(
+      resourceTypes: Map[String, String],
+      resource: Resource): Resource = {
+    require(resource != null, "Resource parameter should not be null!")
+
+    if (!ResourceTypeHelper.isYarnResourceTypesAvailable() && resourceTypes.nonEmpty) {
```

Comment:
do you mean to return whether or `resourceTypes` is empty, but only log the warning if its empty?

I suspect that is the cause of the test failures

Reply:
You are right, this condition was wrong so I added an inner-if to check whether the resourceTypes map is not empty.
The purpose of this check is: Return resource immediately if YARN resource types is not available (Hadoop version is less than 3) and log the warning if YARN resource types is not available and custom resource types are requested.